File: src/main/java/at/irian/webstack/ui/person/PersonDetailView.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package at.irian.webstack.ui.person;

import at.irian.webstack.middle.bo.PersonFacadeLocal;
import at.irian.webstack.middle.orm.Address;
import at.irian.webstack.middle.orm.Person;

import java.io.Serializable;
import javax.inject.Inject;
import javax.inject.Named;

import org.apache.myfaces.extensions.cdi.core.api.scope.conversation.ViewAccessScoped;

/**
 * @author werpu2
 */
@Named
@ViewAccessScoped
public class PersonDetailView implements Serializable {
    public static final String NAV_DETAIL = "personDetail";
    public static final String NAV_LIST = "personList";
    private static final String MODE_CREATE = "create";
    private static final String MODE_EDIT = "edit";
    private static final String MODE_DELETE = "delete";

    @Inject
    PersonListSearchModel searchData;

    @Inject
    PersonFacadeLocal personFacade;

    Person person;
    Address address;

    Long personId = -1L;

    String viewMode = MODE_CREATE;

    public String goCreate() {
        viewMode = MODE_CREATE;
        person = personFacade.create();

        return NAV_DETAIL;
    }

    public String goDeta() {
        viewMode = MODE_EDIT;
        person = personFacade.loadById(personId);

        return NAV_DETAIL;
    }

    public String goDelete() {
        viewMode = MODE_DELETE;
        person = personFacade.loadById(personId);

        return NAV_DETAIL;
    }

    public String doDelete() {
        personFacade.delete(person);

        return NAV_LIST;
    }

    public String doSave() {

        personFacade.save(person);



        return NAV_LIST;
    }

    public String doCancel() {

        personFacade.cancel(person);

        return NAV_LIST;
    }

    public Long getPersonId() {
        return personId;
    }

    public void setPersonId(Long personId) {
        this.personId = personId;
    }

    public Person getPerson() {
        return person;
    }

    public void setPerson(Person person) {
        this.person = person;
    }

    public String getViewMode() {
        return viewMode;
    }

    public void setViewMode(String viewMode) {
        this.viewMode = viewMode;
    }

    public PersonListSearchModel getSearchData() {
        return searchData;
    }

    public void setSearchData(PersonListSearchModel searchData) {
        this.searchData = searchData;
    }

    public Address getAddress() {
        return address;
    }

    public void setAddress(Address address) {
        this.address = address;
    }

    public String addAddress() {
        address = personFacade.createAdr();
        person.getAddresses().add(address);

        return NAV_DETAIL;
    }

    public String removeAddress() {
        person.getAddresses().remove(address);

        return NAV_DETAIL;
    }

    //now to the tricky part, in clustered situations the detail view or the session
    //Generally can be serialised and deserialised in this case we lose the entity manager
    //connection
    //since per spec the em hosting ejb never is passivated thanks to the em injected
    //We simply restore the person bean from there, the altered data still should be present
    //todo deserialisation, we reload the person object anew
    /*private void readObject(ObjectInputStream in) throws IOException, ClassNotFoundException {
        if(personId != -1L) {
            person = personFacade.loadById(personId);
        }
    } */

}

File: web/src/main/scala/org/extrasapache/myfaces/codi/examples/ebean/view/person/PersonDetailView.scala
package org.extrasapache.myfaces.codi.examples.ebean.view.person

import org.apache.myfaces.extensions.cdi.core.api.scope.conversation.ViewAccessScoped
import javax.inject.{Inject, Named}
import org.extrasapache.myfaces.codi.examples.ebean.business.bo.person.PersonFacade
import org.extrasapache.myfaces.codi.examples.ebean.orm.person.Address

object PersConst {
  val MODE_CREATE = "create"
  val MODE_EDIT = "edit"
  val MODE_DELETE = "delete"
}

/**
 * PersonListViewModel
 * we split our page bean into two parts
 * a model part containing all referenced
 * model elements (and combined controller model elements)
 *
 * and a controller class part containing the core logic
 * That way we get a cleaner structure  and can share models
 * between various page beans with different logic
 *
 */
trait PersonDetailViewModel {

  @Inject
  var searchData: PersonListSearchModel = _

  @Inject
  var personFacade: PersonFacade = _

  var person: org.extrasapache.myfaces.codi.examples.ebean.orm.person.Person = _

  var address: Address = _

  var viewMode: String = PersConst.MODE_CREATE
}

/**
 * The controller logic which itself is also a page bean,
 * we use traits as decorators which decorate our model in
 */
@Named
@ViewAccessScoped
@serializable
class PersonDetailView extends PersonDetailViewModel {

  //an import of Person enables GO_DETA and GO_LIST
  import Person._

  def goCreate: java.lang.Class[_] = {
    viewMode = PersConst.MODE_CREATE
    person = personFacade.create

    GO_DETA
  }

  def goDeta: java.lang.Class[_] = {
    viewMode = PersConst.MODE_EDIT

    GO_DETA
  }

  def doDelete: java.lang.Class[_] = {
    personFacade.delete(person)
    GO_LIST
  }

  def doSave: java.lang.Class[_] = {
    personFacade.save(person)
    GO_LIST
  }

  def doCancel: java.lang.Class[_] = {
    personFacade.cancel(person)
    GO_LIST
  }

  def addAddress: java.lang.Class[_] = {
    address = personFacade.createAdr
    person.addresses.add(address)

    GO_DETA
  }

  def removeAddress: java.lang.Class[_] = {
    person.addresses.remove(address)

    GO_DETA
  }

}